wire/msginitialstate.go

```diff
@@ -0,0 +1,289 @@
+// MaxISTSpendsAtHeadPerMsg may be added before this function errors out.
+func (msg *MsgInitState) AddTSpendHash(hash *chainhash.Hash) error {
+	const op = "MsgInitState.AddTSpendHash"
+	if len(msg.TSpendHashes)+1 > MaxISTSpendsAtHeadPerMsg {
+		msg := fmt.Sprintf("too many tspend hashes for message [max %v]",
+			MaxISTSpendsAtHeadPerMsg)
+		return messageError(op, ErrTooManyTSpends, msg)
+	}
+
+	msg.TSpendHashes = append(msg.TSpendHashes, *hash)
+	return nil
+}
+
+// BtcDecode decodes r using the protocol encoding into the receiver.
+// This is part of the Message interface implementation.
+func (msg *MsgInitState) BtcDecode(r io.Reader, pver uint32) error {
+	const op = "MsgInitState.BtcDecode"
+	if pver < InitStateVersion {
+		msg := fmt.Sprintf("%s message invalid for protocol version %d",
+			msg.Command(), pver)
+		return messageError(op, ErrMsgInvalidForPVer, msg)
+	}
+
+	// Read num block hashes and limit to max.
+	count, err := ReadVarInt(r, pver)
+	if err != nil {
+		return err
+	}
+	if count > MaxISBlocksAtHeadPerMsg {
+		msg := fmt.Sprintf("too many block hashes for message "+
+			"[count %v, max %v]", count, MaxISBlocksAtHeadPerMsg)
+		return messageError(op, ErrTooManyBlocks, msg)
+	}
+
+	msg.BlockHashes = make([]chainhash.Hash, count)
+	for i := uint64(0); i < count; i++ {
+		err := readElement(r, &msg.BlockHashes[i])
+		if err != nil {
+			return err
+		}
+	}
+
+	// Read num vote hashes and limit to max.
+	count, err = ReadVarInt(r, pver)
+	if err != nil {
+		return err
+	}
+	if count > MaxISVotesAtHeadPerMsg {
+		msg := fmt.Sprintf("too many vote hashes for message "+
+			"[count %v, max %v]", count, MaxISVotesAtHeadPerMsg)
+		return messageError(op, ErrTooManyVotes, msg)
+	}
+
+	msg.VoteHashes = make([]chainhash.Hash, count)
+	for i := uint64(0); i < count; i++ {
+		err := readElement(r, &msg.VoteHashes[i])
+		if err != nil {
+			return err
+		}
+	}
+
+	// Read num tspend hashes and limit to max.
+	count, err = ReadVarInt(r, pver)
+	if err != nil {
+		return err
+	}
+	if count > MaxISTSpendsAtHeadPerMsg {
+		msg := fmt.Sprintf("too many tspend hashes for message "+
+			"[count %v, max %v]", count, MaxISTSpendsAtHeadPerMsg)
+		return messageError(op, ErrTooManyTSpends, msg)
+	}
+
+	msg.TSpendHashes = make([]chainhash.Hash, count)
+	for i := uint64(0); i < count; i++ {
+		err := readElement(r, &msg.TSpendHashes[i])
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// BtcEncode encodes the receiver to w using the protocol encoding.
+// This is part of the Message interface implementation.
+func (msg *MsgInitState) BtcEncode(w io.Writer, pver uint32) error {
+	const op = "MsgInitState.BtcEncode"
+	if pver < InitStateVersion {
+		msg := fmt.Sprintf("%s message invalid for protocol version %d",
+			msg.Command(), pver)
+		return messageError(op, ErrMsgInvalidForPVer, msg)
+	}
+
+	// Write block hashes.
+	count := len(msg.BlockHashes)
+	if count > MaxISBlocksAtHeadPerMsg {
```

Comment:
i wonder if it wouldn't be better to check all of these max counts before writing anything, to avoid partially writing a message.

It might be a little futile however since there are going to be other existing messages that also suffer this issue (and can't be easily checked up front without extra iterations).

Reply:
Maybe (as a future enhacement) the `Message` interface could have a `ValidToEncode() bool` func that does all this checking so all messages behave appropriately and you can safely encode a set of messages into a contiguous buffer, knowing beforehand whether the entire set will be valid or not.